winlogbeat/sys/wineventlog/renderer.go

```diff
@@ -394,3 +389,117 @@ func (r *Renderer) formatMessageFromTemplate(msgTmpl *template.Template, values
 
 	return string(bb.Bytes()), nil
 }
+
+// XMLRenderer is used for converting event log handles into complete events.
+type XMLRenderer struct {
+	conf          RenderConfig
+	metadataCache *publisherMetadataCache
+	renderBuf     []byte
+	render        func(event EvtHandle, out io.Writer) error // Function for rendering the event to XML.
+
+	log *logp.Logger
+}
+
+// NewXMLRenderer returns a new Renderer.
+func NewXMLRenderer(conf RenderConfig, session EvtHandle, log *logp.Logger) *XMLRenderer {
+	const renderBufferSize = 1 << 19 // 512KB, 256K wide characters
+	rlog := log.Named("xml_renderer")
+	r := &XMLRenderer{
+		conf:          conf,
+		renderBuf:     make([]byte, renderBufferSize),
+		metadataCache: newPublisherMetadataCache(session, conf.Locale, rlog),
+		log:           rlog,
+	}
+	// Forwarded events should be rendered using RenderEventXML. It is more
+	// efficient and does not attempt to use local message files for rendering
+	// the event's message.
+	switch conf.IsForwarded {
+	case true:
+		r.render = func(event EvtHandle, out io.Writer) error {
+			return RenderEventXML(event, r.renderBuf, out)
+		}
+	case false:
+		r.render = func(event EvtHandle, out io.Writer) error {
+			get := func(providerName string) EvtHandle {
+				md, _ := r.metadataCache.getPublisherStore(providerName)
+				if md.Metadata != nil {
+					return md.Metadata.Handle
+				}
+				return NilHandle
+			}
+			return RenderEvent(event, conf.Locale, r.renderBuf, get, out)
+		}
+	}
+	return r
+}
+
+// Close closes all handles held by the Renderer.
+func (r *XMLRenderer) Close() error {
+	if r == nil {
+		return errors.New("closing nil renderer")
+	}
+	return r.metadataCache.close()
+}
+
+// Render renders the event handle into an Event.
+func (r *XMLRenderer) Render(handle EvtHandle) (*winevent.Event, string, error) {
+	// From this point on it will return both the event and any errors. It's
+	// critical to not drop data.
+	var errs []error
+
+	bb := sys.NewPooledByteBuffer()
```

Comment:
why not using `bytes.NewBuffer(r.renderBuf)`?

Reply:
That would use renderBuf to read and write to it at the same time. I did test it and performs way worse. What I did is added a single outBuf to not pick new ones from the pool as it seems unneeded.

This whole code will need some refactoring after 9.0, but now the aim is to reuse the current xml render path to match functionality.